sky/provision/gcp/instance.py

```diff
@@ -244,3 +248,19 @@ def cleanup_ports(
         firewall_rule_name = provider_config['firewall_rule']
         instance_utils.GCPComputeInstance.delete_firewall_rule(
             project_id, firewall_rule_name)
+
+
+def query_ports(
+    cluster_name_on_cloud: str,
+    ip: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> Dict[int, List[common.Endpoint]]:
```

Comment:
I'm a little confused about the return value here. Are we returning all types of supported endpoints here? If that is the case, shall we return the HTTP & HTTPS endpoint too?

Reply:
Yes, we return supported endpoints. For a clouds, since we have access to raw sockets we simply return `common.SocketEndpoint`. We cannot return HTTP & HTTPS endpoint here since we don't know if the user is running a HTTP service on the socket.